examples/llava/clip.cpp

```diff
@@ -846,17 +864,32 @@ static ggml_cgraph * clip_image_build_graph(clip_ctx * ctx, const clip_image_f32
         cur = ggml_add(ctx0, embeddings, cur);
 
         embeddings = cur;
-
     }
 
     // post-layernorm
-    if (ctx->has_post_norm) {
+    if (ctx->has_post_norm && max_feature_layer == n_layer) {
         embeddings = ggml_norm(ctx0, embeddings, eps);
         ggml_set_name(embeddings, "post_ln");
 
         embeddings = ggml_add(ctx0, ggml_mul(ctx0, embeddings, model.post_ln_w), model.post_ln_b);
     }
 
+    // final layer is a vision feature layer
+    for (int vl_idx = 0; vl_idx < MAX_IMAGE_FEATURE_LAYERS && (hparams.vision_feature_layer[vl_idx] > 0); vl_idx++) {
+        if (n_layer == ctx->vision_model.hparams.vision_feature_layer[vl_idx]) {
+            embedding_stack.push_back(embeddings);
+            break;
+        }
+    }
+
+    // If feature layers are explicitly set, stack them (if we have multiple)
+    if (embedding_stack.size() > 0) {
+        embeddings = embedding_stack.at(0);
+        for (unsigned long i=1; i < embedding_stack.size(); i++) {
+            embeddings = ggml_concat(ctx0, embeddings, embedding_stack.at(i), 0);
```

Comment:
Using `ggml_concat` on each embedding is fine for now, but please be aware that it may use a lot more memory.

Another way to do is to manually create a result tensor (with dimension that can hold all the embeddings), then use `ggml_view_*` with appropriate offset.

But this is a minor optimization, probably can be done later if needed.

Reply:
Makes sense, I thought this felt pretty efficient! Thank you for the guidance - I'm happy to submit a follow-up PR making this change in the near future 🙂